README.md
# Custom Hamcrest Matchers

Highly reusable custom hamcrest matchers

## Functions

- assert_that_raises

## Available matchers

-	empty
-	date_iso (ISO 8601 formatted date string)
- 	iterable
-	has_len
-	has_keys
-	matches_re
-	callable_
-	json_
- 	subset_of
- 	superset_of 
- 	disjoint_with

### xml matchers
-	xml_document
-	xml_root
-	xml_element
-	xml_contains_element
-	xml_namespaced
-	soap_document
-	soap_message

## Installation

	python setup.py install

## Dependences

- pyHamcrest


## Documentation

### assert_that_raises

```python
from hamcrest import *
from matchers import assert_that_raises

with assert_that_raises(Warning):
	raise Warning()

with assert_that_raises(instance_of(Warning)):
	raise Warning()

with assert_that_raises(has_property('message', has_string(u'warning'))):
	raise Warning(u'this is a warning')

# this raises AssertionError: no Exception raised
with assert_that_raises(NameError):
	raise Warning()

# {'exception': Warning(u'this is a warning')}
with assert_that_raises(Warning) as captured:
	raise Warning(u'this is a warning')

print captured['exception']
```

### empty

```python
from hamcrest import *
from matchers import empty

assert_that(str(), is_(empty()))
assert_that(set(), is_(empty()))
assert_that(dict(), is_(empty()))
assert_that(list(), is_(empty()))
assert_that(tuple(), is_(empty()))
assert_that(unicode(), is_(empty()))
```

It's smart enough to deal with iterators and generators

```python
assert_that(iter([]), is_(empty()))
assert_that((i for i in []), is_(empty()))
```

### date_iso (ISO 8601 formatted date string)

```python
from hamcrest import *
from matchers import date_iso

assert_that('1988-10-04T06:15:00.230943Z', is_(date_iso()))
```

### iterable

```python
from hamcrest import *
from matchers import iterable


assert_that(list(), is_(iterable()))
assert_that(dict(), is_(iterable()))
assert_that(tuple(), is_(iterable()))
assert_that(set(), is_(iterable()))

assert_that(str(), is_(iterable()))
assert_that(unicode(), is_(iterable()))

assert_that((i for i in []), is_(iterable()))
assert_that(iter([]), is_(iterable()))

class IterateMe(object):
	l = list()
	def __iter__(self):
		return iter(l)

assert_that(IterateMe(), is_(iterable()))
```


### has_len

Reimplementation of has_length made to work with generators as well

```python
from hamcrest import *
from matchers import has_len

assert_that((i for i in [1, 2, 3]), has_len(3))
```


### has_keys

```python
from hamcrest import *
from matchers import has_keys

dictionary = {
	'first': 'foo',
	'second': 'bar'
}

assert_that(dictionary, has_keys(['first', 'second']))
```

### matches_re

```python
from hamcrest import *
from matchers import matches_re

assert_that('pattern', matches_re(r'pattern'))
```

### callable_

```python
from hamcrest import *
from matchers import callable_

assert_that(lambda : 'foo', is_(callable_()))
```

### json_

```python
from hamcrest import *
from matchers import json_

assert_that("{'foo': ['bar']}", is_(json_()))
assert_that("{'foo': ['bar']}", is_(json_(has_key('foo'))))
```

### subset_of

```python
from hamcrest import *
from matchers import subset_of

assert_that([1, 2], is_(subset_of([1, 2, 3])))
```

### superset_of

```python
from hamcrest import *
from matchers import superset_of

assert_that([1, 2, 3], is_(superset_of([1, 2])))
```

### disjoint_with

```python
from hamcrest import *
from matchers import disjoint_with

assert_that([1, 2, 3], is_(disjoint_with([4, 5, 6])))
```

### xml_document

```python
from hamcrest import *
from matchers import xml_document
from xml.etree import Element

assert_that('<element/>', is_(xml_document()))
assert_that('<element/>', is_(xml_document(instance_of(Element))))
```

### xml_root

```python
from hamcrest import *
from matchers import xml_root

assert_that('<element/>', xml_root(tag='element'))
```

### xml_element

```python
from hamcrest import *
from matchers import xml_document, xml_element

assert_that('<element/>', is_(xml_element('element')))
assert_that('<element/>', is_(xml_element('element', another_matcher)))
assert_that('<foo:element/>', is_(xml_element(tag='element', ns='foo')))
```

### xml_contains_element

```python
from hamcrest import *
from matchers import xml_root, xml_element, xml_contains_element

assert_that('<parent><child/></parent>', 
	is_(xml_element('parent', xml_contains_element('child'))))

assert_that('<parent><child/></parent>', 
	xml_root(is_(xml_element('parent', xml_contains_element('child')))))
```

### xml_namespaced

```python
from hamcrest import *
from matchers import xml_namespaced

assert_that('<element xmlns="http://foo.com"/>',
	is_(xml_namespaced('http://foo.com')))
```

### soap_document

```python
from hamcrest import *
from matchers import xml_document, soap_document

ns_url = "http://schemas.xmlsoap.org/soap/envelope/"
string = "<Envelope xmlns='" + ns_url + "' />"

assert_that(string, is_(xml_document(is_(soap_document()))))
```

### soap_message

```python
from hamcrest import *
from matchers import xml_document, soap_document, soap_message

ns_url = "http://schemas.xmlsoap.org/soap/envelope/"
string = """
	<Envelope xmlns='""" + ns_url + """' >"
		<Body/>
	</Envelope>
"""

assert_that(string, 
		is_(xml_document(is_(soap_document(is_(soap_message()))))))
```


CHANGES.txt
0.22  (2013-03-12)
------------------

* Añade primer matcher is_displayed para tests de selenium

0.21  (2013-03-04)
------------------

* Elimina lxml como dependencia

0.2  (2013-03-04)
-----------------

* Mejora empaquetado


0.1.9 (2013-03-04)
------------------

* Añade matcher has_len
* Elimina matcher has_properties que aparece en PyHamcrest==1.7.1
* Añade versión reStructuredText del README


0.1.8 (2013-02-28)
------------------

* Añade nueva función assert_that_raises
* Añade nuevo matcher subset_of
* Añade nuevo matcher superset_of
* Añade nuevo matcher disjoint_with
* Reimplementa matcher 'empty'
* Arregla cadenas unicode sueltas

0.1.7 (2013-02-18)
------------------

* Añade nuevo matcher 'iterable'


0.1.6 (2013-02-18)
------------------

* Añade uso de *args a has_keys


0.1.5 (2013-02-18)
------------------

* Añade uso de kwargs a has_properties
* Añade tests para has_properties


tests/test_xml.py
# -*- coding: utf-8 -*-
"""
This file is part of matchers.

matchers is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

matchers is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with matchers.  If not, see <http://www.gnu.org/licenses/>.
"""

from hamcrest import *

from xml.etree import ElementTree

from matchers import xml_document


class TestXMLDocument(object):
    document = """
    <element xmlns="http://espacio.com">
        <top_level>
            <first_inner_level>
                <second_inner_level>
                    <empty_element/>
                </second_inner_level>
            </first_inner_level>
        </top_level>
    </element>
    """

    document_consistent = "<?xml version='1.0' ?>" + document
    document_inconsistent = "<?xml version='1.0' encoding='utf-8'?>" + document


    def setup(self):
        self.dom = ElementTree.fromstring(self.document)

    def test_xml_document(self):
        assert_that(self.document, is_(xml_document()))

    def test_xml_document_extra_matcher_works(self):
        assert_that(self.document, is_(xml_document(is_(instance_of(type(self.dom))))))

    def test_xml_document_encoding_inconsistence_does_not_trigger_exceptions(self):
        assert_that(self.document_inconsistent, is_(xml_document()))

    def test_xml_document_encoding_consistency_just_works(self):
        assert_that(self.document_consistent, is_(xml_document()))


tests/test_empty.py
# -*- coding: utf-8 -*-
"""
This file is part of matchers.

matchers is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

matchers is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with matchers.  If not, see <http://www.gnu.org/licenses/>.
"""

from hamcrest import *

from matchers import empty


class TestEmpty(object):
    def test_empty_dict(self):
        assert_that(dict(), is_(empty()))

    def test_empty_tuple(self):
        assert_that(tuple(), is_(empty()))

    def test_empty_list(self):
        assert_that(list(), is_(empty()))

    def test_empty_set(self):
        assert_that(set(), is_(empty()))

    def test_empty_unicode(self):
        assert_that(u'', is_(empty()))

    def test_empty_str(self):
        assert_that(str(), is_(empty()))

    def test_empty_generator(self):
        assert_that((a for a in list()), is_(empty()))

    def test_empty_iterator(self):
        assert_that(iter(list()), is_(empty()))


class TestNotEmpty(object):
    def setup(self):
        self.tuple = ((1, 2), (3, 4))

    def test_not_empty_dict(self):
        assert_that(dict(self.tuple), is_not(empty()))

    def test_not_empty_tuple(self):
        assert_that(self.tuple, is_not(empty()))

    def test_not_empty_list(self):
        assert_that(list(self.tuple), is_not(empty()))

    def test_not_empty_set(self):
        assert_that(set(self.tuple), is_not(empty()))

    def test_not_empty_unicode(self):
        assert_that(u'unicode', is_not(empty()))

    def test_not_empty_str(self):
        assert_that('string', is_not(empty()))

    def test_not_empty_generator(self):
        assert_that((i for i in range(1)), is_not(empty()))

    def test_not_empty_iterator(self):
        assert_that(iter([1]), is_not(empty()))


tests/test_regex.py
# -*- coding: utf-8 -*-
"""
This file is part of matchers.

matchers is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

matchers is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with matchers.  If not, see <http://www.gnu.org/licenses/>.
"""

from hamcrest import *

from matchers import matches_re, date_iso


DATE_ISO_FORMAT = '2012-10-10T00:00:00'
DATE_ISO_FORMAT_SPACE = '2012-10-10 00:00:00'
DATE_ISO_FORMAT_SHORT = '2012-10-10'
DATE_ISO_FORMAT_MILISECONDS = '2012-10-10T00:00:00.2042342'
DATE_ISO_FORMAT_WRONG_ORDER = '10-2012-10'
DATE_ISO_FORMAT_WRONG_ORDER_1 = '21-10-2012'
DATE_ISO_FORMAT_WRONG_MONTH = '2012-13-10T00:00:00'
DATE_ISO_FORMAT_WRONG_DAY = '2012-10-40T00:00:00'
DATE_ISO_FORMAT_WRONG_HOUR = '2012-10-10T25:00:00'
DATE_ISO_FORMAT_WRONG_MIN = '2012-10-10T00:70:00'
DATE_ISO_FORMAT_WRONG_SEC = '2012-10-10T00:00:70'


class TestMatchesRe(object):
    def test_matches_text(self):
        assert_that('test', matches_re('test'))

    def test_finds_text(self):
        assert_that('...test....', matches_re('test'))

    def test_matches_regex(self):
        assert_that('ACAB', matches_re('[AB][BC][AB][BC]'))

    def test_finds_regex(self):
        assert_that('....ACAB....', matches_re('[AB][BC][AB][BC]'))


class TestMatchesDateISO(object):
    def test_matches_iso_date(self):
        assert_that(DATE_ISO_FORMAT, is_(date_iso()))
        assert_that(DATE_ISO_FORMAT_SPACE, is_(date_iso()))

    def test_matches_iso_date_with_miliseconds(self):
        assert_that(DATE_ISO_FORMAT_MILISECONDS, is_(date_iso()))

    def test_matchers_iso_date_without_time(self):
        assert_that(DATE_ISO_FORMAT.split('T')[0], is_(date_iso()))

    def test_matchers_iso_date_with_slash_separator(self):
        assert_that(DATE_ISO_FORMAT.replace('-', '/'), is_(date_iso()))

    def test_matchers_iso_date_with_different_order(self):
        assert_that(DATE_ISO_FORMAT_WRONG_ORDER, is_not(date_iso()))
        assert_that(DATE_ISO_FORMAT_WRONG_ORDER_1, is_not(date_iso()))

    def test_matchers_iso_date_with_wrong_month(self):
        assert_that(DATE_ISO_FORMAT_WRONG_MONTH, is_not(date_iso()))

    def test_matchers_iso_date_with_wrong_day(self):
        assert_that(DATE_ISO_FORMAT_WRONG_DAY, is_not(date_iso()))

    def test_matchers_iso_date_with_wrong_hour(self):
        assert_that(DATE_ISO_FORMAT_WRONG_HOUR, is_not(date_iso()))

    def test_matchers_iso_date_with_wrong_min(self):
        assert_that(DATE_ISO_FORMAT_WRONG_MIN, is_not(date_iso()))

    def test_matchers_iso_date_with_wrong_sec(self):
        assert_that(DATE_ISO_FORMAT_WRONG_SEC, is_not(date_iso()))


matchers/__init__.py
# -*- coding: utf-8 -*-
"""
This file is part of matchers.

matchers is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

matchers is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with matchers.  If not, see <http://www.gnu.org/licenses/>.
"""

import re
import json
from xml.etree import ElementTree
from collections import Iterable
from contextlib import contextmanager

from hamcrest import *
from hamcrest.core.base_matcher import BaseMatcher


class matches_re(BaseMatcher):
    """Custom hamcrest matcher for regular expressions"""

    def __init__(self, regex):
        self.regex_text = regex
        self.regex = re.compile(regex)

    def _matches(self, item):
        if not isinstance(item, str) and not isinstance(item, unicode):
            return False

        return any(self.regex.finditer(item))

    def describe_to(self, description):
        description.append_text(u" string that matches regex: '{}'".format(self.regex_text))

    def describe_mismatch(self, actual, description):
        description.append_text(u" string '{}' does not match regex: '{}'"
                                .format(actual, self.regex_text))


class edi_document(matches_re):
    """Custom hamcrest matcher to pseudo-validate EDI"""
    def __init__(self):
        super(edi_document, self).__init__(
            u'UNB\+UNOA:1\+[^:]+:ZZ\+AEATADUE:ZZ\+\d{8}:\d{4}\+\d+\+\+&EE'
        )


class date_iso(matches_re):
    fmt = dict(
        year=r'(\d{4})',
        month=r'((0[1-9])|(1[0-2]))',
        day=r'([0-3]\d)',
        hours=r'(([01]\d)|(2[0-3]))',
        minutes=r'([0-5]\d)',
        seconds=r'([0-5]\d)',
        miliseconds=r'(\d+)',
        sep=r'[/-]',
    )

    def __init__(self):
        "1988-10-04T06:15:00.230943Z"
        super(date_iso, self).__init__(
            u'^{year}{sep}{month}{sep}{day}([T ]{hours}:{minutes}:{seconds}(.{miliseconds})?)?$'
            .format(**self.fmt)
        )

    def describe_to(self, description):
        description.append_text(u' an ISO 8601 formatted date string ')


class xml_document(BaseMatcher):
    """Custom hamcrest matcher for XML documents"""
    def __init__(self, matcher=None):
        self.matcher = matcher

    def _matches(self, item):
        try:
            try:
                doc = ElementTree.fromstring(item)
            except ValueError as err:
                # try again at ValueError: Unicode strings with encoding declaration are not supported.
                item = item.encode('utf-8')
                doc = ElementTree.fromstring(item)
        except Exception as err:
            self.error = err

        if self.matcher:
            return self.matcher.matches(doc)

        return True

    def describe_to(self, description):
        description.append_text(u"string conforming a valid XML document")

    def describe_mismatch(self, actual, description):
        description.append_text(u"failed to load xml document {} due to following errors: '{}'"
                                .format(actual, self.error))


class xml_element(BaseMatcher):
    """Checks for a xml element with a certain name"""
    def __init__(self, tag, matcher=None, ns=None):
        self.tag = tag if ns is None else "{{{ns}}}{tag}".format(ns=ns, tag=tag)
        self.matcher = matcher

    def _matches(self, item):
        if not item.tag == self.tag:
            return False

        if self.matcher:
            return self.matcher.matches(item)

        return True

    def describe_to(self, description):
        description.append_text(u" xml element with tag named '{0}'".format(self.tag))


# A etree xml document is the root element also.
# This may change in other implementation.
xml_root = xml_element


class xml_contains_element(xml_element):
    def _matches(self, item):
        try:
            item = item.find(".//" + self.tag)
        except:
            return False

        if item is None:
            return False

        return super(xml_contains_element, self)._matches(item)

    def describe_to(self, description):
        description.append_text(u" xml element has a children: '{0}'".format(self.tag))
        if self.matcher:
            description.append_text(u" matching {}".format(self.matcher))


class xml_namespaced(BaseMatcher):
    def __init__(self, ns_url):
        self.url = ns_url

    def _matches(self, item):
        return self.url in item.nsmap.values()

    def describe_to(self, description):
        description.append_text(u"xml document wich is namespaced with the url '{0}'".format(self.url))

    def describe_mismatch(self, actual, description):
        description.append_text(u"url '{}' is not within the namespaces of the xml element {}".format(self.url, actual.nsmap))


class soap_document(xml_document):
    """Checks for a well constructed soap document. xml_document with extra checking"""

    ns_url = "http://schemas.xmlsoap.org/soap/envelope/"

    def __init__(self, matcher=None):
        default_matcher = all_of(xml_root('Envelope', ns=self.ns_url))
        if matcher:
            default_matcher = all_of(default_matcher, matcher)

        super(soap_document, self).__init__(default_matcher)

    def describe_to(self, description):
        description.append_text(u"SOA xml document envelope")

    def describe_mismatch(self, actual, description):
        description.append_text(u"is not a valid xml document or does not have namespace '{0}'"
                                .format(self.ns_url))


class soap_message(soap_document):
    def __init__(self, matcher=None):
        super(soap_message, self).__init__(xml_contains_element('Body', ns=self.ns_url, matcher=matcher))

    def describe_to(self, description):
        description.append_text(u"xml document within a SOA envelope")

    def describe_mismatch(self, actual, description):
        description.append_text(u" soap message {} should containg a 'Body' element ".format(actual))


class json_(BaseMatcher):
    def __init__(self, matcher=None):
        self.matcher = matcher
        self.response = None

    def _matches(self, response):
        self.response = response
        try:
            self.response = json.loads(response)
        except:
            return False

        if self.matcher:
            return self.matcher.matches(self.response)

        return True

    def describe_to(self, description):
        description.append_text(u'a JSON object that ').append_description_of(self.matcher)

    def describe_mismatch(self, response, description):
        description.append_text(u'was ').append_value(self.response)
        if self.matcher:
            self.matcher.describe_mismatch(self.response, description)


class callable_(BaseMatcher):
    def _matches(self, item):
        return callable(item)

    def describe_to(self, description):
        description.append_text(u'callable object')

    def describe_mismatch(self, actual, description):
        description.append_text(u' the object {} was not a callable object'.format(actual))


class empty(BaseMatcher):
    def _matches(self, obj):
        try:
            it = iter(obj)
            next = getattr(it, 'next', getattr(it, '__next__', None))
            next()
        except TypeError:
            return not bool(obj)
        except StopIteration:
            return True
        else:
            return False

    def describe_to(self, description):
        description.append_text(u' an empty object ')


class has_keys(BaseMatcher):
    def __init__(self, *args):
        if len(args) == 1 and isinstance(args[0], Iterable):
            args = args[0]

        self.entries = set(args)

    def _matches(self, obj):
        return not bool(self.entries - set(dict(obj).keys()))

    def describe_to(self, description):
        description.append_text(u'a dictionary contains all this keys: {}'.format(self.entries))

    def describe_mismatch(self, actual, description):
        description.append_text(u' but found instead: {}'.format(actual))
        description.append_text(u' which misses keys: {}'.format(self.entries - set(actual.keys())))


class iterable(BaseMatcher):
    def _matches(self, obj):
        return isinstance(obj, Iterable)

    def describe_to(self, description):
        description.append_text(u'an iterable object ')

    def describe_mismatch(self, actual, description):
        description.append_text(u' but found instead a {} object, which is not iterable'.format(actual))


class _setmatcher(BaseMatcher):
    def __init__(self, sequence):
        self.sequence = set(sequence)

    def describe_mismatch(self, actual, description):
        description.append_text(
            u' but found instead {other}, wich differs of the target in {diff}'
            .format(other=self.other_sequence,
                    diff=self.sequence.symmetric_difference(self.other_sequence)))


class subset_of(_setmatcher):
    def _matches(self, other_sequence):
        self.other_sequence = set(other_sequence)
        return self.other_sequence.issubset(self.sequence)

    def describe_to(self, description):
        description.append_text(u'a set of which {} is a subset'.format(self.sequence))


class superset_of(_setmatcher):
    def _matches(self, other_sequence):
        self.other_sequence = set(other_sequence)
        return self.other_sequence.issuperset(self.sequence)

    def describe_to(self, description):
        description.append_text('a set of which {} is superset'.format(self.sequence))


class disjoint_with(_setmatcher):
    def _matches(self, other_sequence):
        self.other_sequence = set(other_sequence)
        return self.other_sequence.isdisjoint(self.sequence)

    def describe_to(self, description):
        description.append_text('a set disjoint of {}'.format(self.sequence))


class as_unicode(BaseMatcher):
    def __init__(self, matcher=None):
        self.matcher = matcher

    def _matches(self, obj):
        try:
            uobj = unicode(obj)
        except UnicodeDecodeError:
            raise AssertionError(u'object "{}" is not unicode codificable'.format(obj))
        else:
            return matcher.matches(unicode(uobj))

    def describe_to(self, description):
        description.append_text(u'an unicode-codificable object')


class has_len(BaseMatcher):
    def __init__(self, length):
        self.length = length

    def _matches(self, obj):
        try:
            length = len(obj)
        except TypeError:
            length = len(list(obj))

        return self.length == length

    def describe_to(self, description):
        description.append_text(u'a sequence of length {}'.format(self.length))


class all_matches(BaseMatcher):
    def __init__(self, matcher):
        self.matcher = matcher

    def _matches(self, sequence):
        match_fun = self.matcher.matches
        return all(imap(match_fun, sequence))

    def describe_to(self, description):
        description.append_text(u'a sequence whose elements all matches: ')
        self.matcher.describe_to(description)


@contextmanager
def assert_that_raises(matcher_or_exception):
    exception, matcher = ((matcher_or_exception, None)
                          if isinstance(matcher_or_exception, Exception)
                          else (Exception, matcher_or_exception))

    catched = {'exception': None}
    try:
        yield catched
    except exception as raised:
        catched['exception'] = raised
        if matcher is not None:
            assert_that(raised, matcher)
    else:
        raise AssertionError(u'no {} raised'.format(exception.__name__))


class is_displayed(BaseMatcher):
    def _matches(self, selenium_element):
        return selenium_element.is_displayed()

    def describe_to(self, description):
        description.append_text(u'a web element which is being displayed')

    def describe_mismatch(self, actual, description):
        description.append_text(u'a non displayed element at position {}'.format(actual.location))
